src/coap_tinydtls.c

```diff
@@ -562,11 +698,638 @@ coap_get_tls_library_version(void) {
   return &version;
 }
 
+#ifdef DTLS_ECC
+static const uint8_t b64_6[256] =
+  {
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+/*                                             +               / */
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63,
+/* 0   1   2   3   4   5   6   7   8   9               =         */
+  52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64,
+/*     A   B   C   D   E   F   G   H   I   J   K   L   M   N   O */
+  64,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
+/* P   Q   R   S   T   U   V   W   X   Y   Z                     */
+  15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64,
+/*     a   b   c   d   e   f   g   h   i   j   k   l   m   n   o */
+  64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
+/* p   q   r   s   t   u   v   w   x   y   z                     */
+  41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+  };
+
+/* caller must free off returned coap_binary_t* */
+static coap_binary_t *
+pem_base64_decode (const uint8_t *data, size_t size)
+{
+  uint8_t *tbuf = coap_malloc(size);
+  size_t nbytesdecoded;
+  size_t i;
+  coap_binary_t *decoded;
+  uint8_t *ptr;
+  uint8_t *out;
+  size_t nb64bytes = 0;
+
+  for (i = 0; i < size; i++) {
+    switch (data[i]) {
+    case ' ':
+    case '\r':
+    case '\n':
+    case '\t':
+      break;
+    default:
+      if (b64_6[data[i]] == 64)
+        goto end;
+      tbuf[nb64bytes++] = data[i];
+      break;
+    }
+  }
+
+end:
+  nbytesdecoded = ((nb64bytes + 3) / 4) * 3;
+  decoded = coap_new_binary(nbytesdecoded + 1);
+  if (!decoded)
+    return NULL;
+
+  out = decoded->s;
+  ptr = tbuf;
+
+  while (nb64bytes > 4) {
+    *(out++) = b64_6[ptr[0]] << 2 | b64_6[ptr[1]] >> 4;
+    *(out++) = b64_6[ptr[1]] << 4 | b64_6[ptr[2]] >> 2;
+    *(out++) = b64_6[ptr[2]] << 6 | b64_6[ptr[3]];
+    ptr += 4;
+    nb64bytes -= 4;
+  }
+
+  /* Note: (nb64bytes == 1) is an error */
+  if (nb64bytes > 1) {
+    *(out++) = b64_6[ptr[0]] << 2 | b64_6[ptr[1]] >> 4;
+  }
+  if (nb64bytes > 2) {
+    *(out++) = b64_6[ptr[1]] << 4 | b64_6[ptr[2]] >> 2;
+  }
+  if (nb64bytes > 3) {
+    *(out++) = b64_6[ptr[2]] << 6 | b64_6[ptr[3]];
+  }
+
+  decoded->length = nbytesdecoded - ((4 - nb64bytes) & 3);
+  coap_free(tbuf);
+  return decoded;
+}
+
+static size_t
+asn1_len(const uint8_t **ptr)
```

Comment:
There is a lot of duplicate code here with the ASN.1 functions in coap_gnutls.c. Maybe we can put them into a separate file and share between both modules?

Reply:
Common code now created for the asn1 functions.